File: mm-swift/Sources/MetaMessage/JSONC/JSONCScanner.swift
import Foundation

public enum JSONCTokenType: Equatable {
    case eof
    case lBrace
    case rBrace
    case lBracket
    case rBracket
    case colon
    case comma
    case string
    case number
    case trueValue
    case falseValue
    case nullValue
    case leadingComment
    case trailingComment
}

public struct JSONCToken {
    public let type: JSONCTokenType
    public let literal: String
    public let line: Int
    public let column: Int

    public init(type: JSONCTokenType, literal: String = "", line: Int, column: Int) {
        self.type = type
        self.literal = literal
        self.line = line
        self.column = column
    }
}

public class JSONCScanner {
    private var src: [Character]
    private var pos: Int
    private var line: Int
    private var col: Int
    private var newLine: Bool

    public init(input: String) {
        self.src = Array(input)
        self.pos = 0
        self.line = 1
        self.col = 1
        self.newLine = false
    }

    private func peek() -> Character {
        guard pos < src.count else { return "\0" }
        return src[pos]
    }

    private func next() -> Character {
        guard pos < src.count else { return "\0" }
        let ch = src[pos]
        pos += 1

        if ch == "\n" {
            newLine = true
            line += 1
            col = 1
        } else {
            col += 1
        }
        return ch
    }

    private func skipWhitespace() {
        while let ch = peekIfNotEmpty(), ch.isWhitespace {
            _ = next()
        }
    }

    private func peekIfNotEmpty() -> Character? {
        guard pos < src.count else { return nil }
        return src[pos]
    }

    public func nextToken() -> JSONCToken {
        skipWhitespace()

        let ch = peekIfNotEmpty()
        if ch == nil || ch == "\0" {
            return JSONCToken(type: .eof, line: line, column: col)
        }

        let startLine = line
        let startCol = col

        switch ch {
        case "{":
            _ = next()
            return JSONCToken(type: .lBrace, line: startLine, column: startCol)
        case "}":
            _ = next()
            return JSONCToken(type: .rBrace, line: startLine, column: startCol)
        case "[":
            _ = next()
            return JSONCToken(type: .lBracket, line: startLine, column: startCol)
        case "]":
            _ = next()
            return JSONCToken(type: .rBracket, line: startLine, column: startCol)
        case ":":
            _ = next()
            newLine = false
            return JSONCToken(type: .colon, line: startLine, column: startCol)
        case ",":
            _ = next()
            newLine = false
            return JSONCToken(type: .comma, line: startLine, column: startCol)
        case "\"":
            return scanString(startLine: startLine, startCol: startCol)
        case "/":
            return scanComment(startLine: startLine, startCol: startCol)
        default:
            return scanLiteral(startLine: startLine, startCol: startCol)
        }
    }

    private func scanString(startLine: Int, startCol: Int) -> JSONCToken {
        _ = next()

        var buf = ""
        while let ch = peekIfNotEmpty(), ch != "\0" {
            let c = next()
            if c == "\n" || c == "\0" {
                break
            }
            if c == "\"" {
                break
            }
            if c == "\\" {
                buf.append(c)
                if let escaped = peekIfNotEmpty(), escaped != "\0" {
                    buf.append(next())
                }
                continue
            }
            buf.append(c)
        }

        return JSONCToken(type: .string, literal: buf, line: startLine, column: startCol)
    }

    private func scanComment(startLine: Int, startCol: Int) -> JSONCToken {
        _ = next()

        if peekIfNotEmpty() == "/" {
            let commentType: JSONCTokenType = newLine ? .leadingComment : .trailingComment
            _ = next()

            var buf = ""
            while let ch = peekIfNotEmpty(), ch != "\0", ch != "\n" {
                buf.append(next())
            }
            let trimmed = buf.trimmingCharacters(in: .whitespaces)

            return JSONCToken(type: commentType, literal: trimmed, line: startLine, column: startCol)
        }

        if peekIfNotEmpty() == "*" {
            let commentType: JSONCTokenType = newLine ? .leadingComment : .trailingComment
            _ = next()

            var buf = ""
            while let ch = peekIfNotEmpty(), ch != "\0" {
                if ch == "*" && pos + 1 < src.count && src[pos + 1] == "/" {
                    _ = next()
                    _ = next()
                    break
                }
                buf.append(next())
            }
            let trimmed = buf.trimmingCharacters(in: .whitespaces)

            return JSONCToken(type: commentType, literal: trimmed, line: startLine, column: startCol)
        }

        return JSONCToken(type: .eof, line: startLine, column: startCol)
    }

    private func scanLiteral(startLine: Int, startCol: Int) -> JSONCToken {
        var buf = ""

        while let ch = peekIfNotEmpty(), ch != "\0", !ch.isWhitespace, !isDelimiter(ch) {
            buf.append(next())
        }

        switch buf {
        case "true":
            return JSONCToken(type: .trueValue, literal: buf, line: startLine, column: startCol)
        case "false":
            return JSONCToken(type: .falseValue, literal: buf, line: startLine, column: startCol)
        case "null":
            return JSONCToken(type: .nullValue, literal: buf, line: startLine, column: startCol)
        default:
            return JSONCToken(type: .number, literal: buf, line: startLine, column: startCol)
        }
    }

    private func isDelimiter(_ ch: Character) -> Bool {
        return ch == "{" || ch == "}" || ch == "[" || ch == "]" || ch == ":" || ch == ","
    }
}

File: mm-cs/src/MetaMessage/Jsonc/JsoncScanner.cs
namespace MetaMessage.Jsonc;

public class JsoncScanner
{
    private readonly string _input;
    private int _position;
    private int _line;
    private int _column;
    private bool _newLine;

    public JsoncScanner(string input)
    {
        _input = input;
        _position = 0;
        _line = 1;
        _column = 1;
    }

    public JsoncToken NextToken()
    {
        SkipWhitespace();
        if (_position >= _input.Length)
        {
            return new JsoncToken
            {
                Type = JsoncTokenType.EOF,
                Line = _line,
                Column = _column
            };
        }

        char ch = _input[_position];

        if (ch == '/')
        {
            return ScanComment();
        }

        switch (ch)
        {
            case '{':
                Advance(1);
                return NewToken(JsoncTokenType.LBrace);
            case '}':
                Advance(1);
                return NewToken(JsoncTokenType.RBrace);
            case '[':
                Advance(1);
                return NewToken(JsoncTokenType.LBracket);
            case ']':
                Advance(1);
                return NewToken(JsoncTokenType.RBracket);
            case ':':
                Advance(1);
                _newLine = false;
                return NewToken(JsoncTokenType.Colon);
            case ',':
                Advance(1);
                _newLine = false;
                return NewToken(JsoncTokenType.Comma);
            case '"':
                return ScanString();
            default:
                if (IsDigit(ch) || ch == '-')
                {
                    return ScanNumber();
                }
                if (IsLetter(ch))
                {
                    return ScanIdentifier();
                }
                throw new Exception($"Unexpected character: {ch} at line {_line}, column {_column}");
        }
    }

    private JsoncToken NewToken(JsoncTokenType type, string literal = "")
    {
        return new JsoncToken
        {
            Type = type,
            Literal = literal,
            Line = _line,
            Column = _column
        };
    }

    private JsoncToken ScanComment()
    {
        if (_input.Length <= _position + 1)
        {
            Advance(1);
            return NewToken(JsoncTokenType.LeadingComment, "/");
        }

        char next = _input[_position + 1];
        int startLine = _line;
        int startColumn = _column;

        if (next == '/')
        {
            Advance(2);
            int contentStart = _position;
            while (_position < _input.Length && _input[_position] != '\n')
            {
                Advance(1);
            }
            string content = _input.Substring(contentStart, _position - contentStart).Trim();
            return new JsoncToken
            {
                Type = _newLine ? JsoncTokenType.LeadingComment : JsoncTokenType.TrailingComment,
                Literal = content,
                Line = startLine,
                Column = startColumn
            };
        }

        if (next == '*')
        {
            Advance(2);
            int contentStart = _position;
            while (_position < _input.Length - 1)
            {
                if (_input[_position] == '*' && _input[_position + 1] == '/')
                {
                    Advance(2);
                    break;
                }
                if (_input[_position] == '\n')
                {
                    _line++;
                    _column = 1;
                }
                Advance(1);
            }
            string content = _input.Substring(contentStart, _position - contentStart - 2).Trim();
            return new JsoncToken
            {
                Type = _newLine ? JsoncTokenType.LeadingComment : JsoncTokenType.TrailingComment,
                Literal = content,
                IsBlock = true,
                Line = startLine,
                Column = startColumn
            };
        }

        Advance(1);
        return NewToken(JsoncTokenType.LeadingComment, "/");
    }

    private JsoncToken ScanString()
    {
        int startLine = _line;
        int startColumn = _column;
        Advance(1);
        var sb = new System.Text.StringBuilder();
        while (_position < _input.Length && _input[_position] != '"')
        {
            if (_input[_position] == '\\' && _position + 1 < _input.Length)
            {
                sb.Append('\\');
                Advance(1);
                sb.Append(_input[_position]);
            }
            else
            {
                if (_input[_position] == '\n')
                {
                    _line++;
                    _column = 1;
                }
                sb.Append(_input[_position]);
            }
            Advance(1);
        }

        if (_position < _input.Length)
        {
            Advance(1);
        }

        return new JsoncToken
        {
            Type = JsoncTokenType.String,
            Literal = sb.ToString(),
            Line = startLine,
            Column = startColumn
        };
    }

    private JsoncToken ScanNumber()
    {
        int startLine = _line;
        int startColumn = _column;
        var sb = new System.Text.StringBuilder();

        if (_input[_position] == '-')
        {
            sb.Append('-');
            Advance(1);
        }

        while (_position < _input.Length && (IsDigit(_input[_position]) || _input[_position] == '.' ||
               _input[_position] == 'e' || _input[_position] == 'E' ||
               _input[_position] == '+' || _input[_position] == '_'))
        {
            if (_input[_position] == '_')
            {
                Advance(1);
                continue;
            }
            sb.Append(_input[_position]);
            Advance(1);
        }

        return new JsoncToken
        {
            Type = JsoncTokenType.Number,
            Literal = sb.ToString(),
            Line = startLine,
            Column = startColumn
        };
    }

    private JsoncToken ScanIdentifier()
    {
        int startLine = _line;
        int startColumn = _column;
        var sb = new System.Text.StringBuilder();

        while (_position < _input.Length && (IsLetter(_input[_position]) || IsDigit(_input[_position])))
        {
            sb.Append(_input[_position]);
            Advance(1);
        }

        string identifier = sb.ToString();
        return identifier.ToLower() switch
        {
            "true" => new JsoncToken { Type = JsoncTokenType.True, Literal = "true", Line = startLine, Column = startColumn },
            "false" => new JsoncToken { Type = JsoncTokenType.False, Literal = "false", Line = startLine, Column = startColumn },
            "null" => new JsoncToken { Type = JsoncTokenType.Null, Literal = "null", Line = startLine, Column = startColumn },
            _ => new JsoncToken { Type = JsoncTokenType.String, Literal = identifier, Line = startLine, Column = startColumn }
        };
    }

    private void SkipWhitespace()
    {
        while (_position < _input.Length && IsWhitespace(_input[_position]))
        {
            if (_input[_position] == '\n')
            {
                _newLine = true;
                _line++;
                _column = 1;
            }
            Advance(1);
        }
    }

    private void Advance(int count)
    {
        _position += count;
        _column += count;
    }

    private static bool IsWhitespace(char ch)
    {
        return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r';
    }

    private static bool IsDigit(char ch)
    {
        return ch >= '0' && ch <= '9';
    }

    private static bool IsLetter(char ch)
    {
        return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
    }
}